lib/Basics/FutureSharedLock.h

```diff
@@ -0,0 +1,246 @@
+
+    void unlock() noexcept {
+      TRI_ASSERT(_lock != nullptr);
+      _lock->unlock();
+      _lock = nullptr;
+    }
+
+    ~LockGuard() {
+      if (_lock) {
+        _lock->unlock();
+      }
+    }
+
+   private:
+    friend struct FutureSharedLock;
+    explicit LockGuard(SharedState* lock) : _lock(lock) {}
+    SharedState* _lock = nullptr;
+  };
+
+  using FutureType = Future<LockGuard>;
+
+  explicit FutureSharedLock(Scheduler& scheduler)
+      : _sharedState(std::make_shared<SharedState>(scheduler)) {}
+
+  FutureType asyncLockShared() {
+    return _sharedState->asyncLockShared([](auto) {});
+  }
+
+  FutureType asyncLockExclusive() {
+    return _sharedState->asyncLockExclusive([](auto) {});
+  }
+
+  FutureType asyncTryLockSharedFor(std::chrono::milliseconds timeout) {
+    return _sharedState->asyncLockShared([this, timeout](auto node) {
+      _sharedState->scheduleTimeout(node, timeout);
+    });
+  }
+
+  FutureType asyncTryLockExclusiveFor(std::chrono::milliseconds timeout) {
+    return _sharedState->asyncLockExclusive([this, timeout](auto node) {
+      _sharedState->scheduleTimeout(node, timeout);
+    });
+  }
+
+  void unlock() { _sharedState->unlock(); }
+
+ private:
+  struct Node {
+    explicit Node(bool exclusive) : exclusive(exclusive) {}
+
+    Promise<LockGuard> promise;
+    bool exclusive;
+    bool scheduled{false};
+  };
+
+  struct SharedState : std::enable_shared_from_this<SharedState> {
+    explicit SharedState(Scheduler& scheduler) : _scheduler(scheduler) {}
+
+    template<class Func>
+    FutureType asyncLockExclusive(Func blockedFunc) {
+      std::lock_guard lock(_mutex);
+      if (_lockCount == 0) {
+        TRI_ASSERT(_queue.empty());
+        ++_lockCount;
+        _exclusive = true;
+        return {LockGuard(this)};
+      }
+
+      auto it = insertNode(true);
+      blockedFunc(it);
+      return (*it)->promise.getFuture();
+    }
+
+    template<class Func>
+    FutureType asyncLockShared(Func blockedFunc) {
+      std::lock_guard lock(_mutex);
+      if (_lockCount == 0 || (!_exclusive && _queue.empty())) {
+        ++_lockCount;
+        _exclusive = false;
+        return {LockGuard(this)};
+      }
+
+      auto it = insertNode(false);
+      blockedFunc(it);
+      return (*it)->promise.getFuture();
+    }
+
+    void unlock() {
+      std::lock_guard lock(_mutex);
+      TRI_ASSERT(_lockCount > 0);
+      if (--_lockCount == 0 && !_queue.empty()) {
+        // we were the last lock holder -> schedule the next node
+        auto& node = _queue.back();
+        _lockCount = 1;
+        _exclusive = node->exclusive;
+        scheduleNode(node);
+        _queue.pop_back();
+
+        // if we are in shared mode, we can schedule all following shared
+        // nodes
+        if (!_exclusive) {
+          while (!_queue.empty() && !_queue.back()->exclusive) {
+            ++_lockCount;
+            scheduleNode(_queue.back());
+            _queue.pop_back();
+          }
```

Comment:
What do you mean? In the "normal" RWLock implementation we always prefer writers. This lock is strictly fair, but we definitely do _not_ want to prefer readers, because this would be a fundamental change compared to the current RWLock implementation.